### PR TITLE
Bug not recovering after an error

### PR DESCRIPTION
**Pull request by EranOfer**
In the producer made each produce request get its own response with a new offset the corrosponds to the request.
1. In the producer made Ack=0 requests return a single response with offset -1
2. In TcpSocket made a finite number of retries. After all retries fail, putting exceptions in each response for each request.
3. In TcpSocket. Removed dependancy between current tasks and future tasks

### CODE REVIEW

File: src/KafkaNetClient/Consumer.cs [hunk 1]
```diff
@@ -124,15 +126,15 @@ private Task ConsumeTopicPartitionAsync(string topic, int partitionId)
...
-                        {
-                            await _options.Router.RefreshTopicMetadata(topic);
-                            EnsurePartitionPollingThreads();
-                            refreshMetaData = false;
-                        }
-
                         try
                         {
+                            //after error
+                            if (refreshMetaData)
```
Comment (yoniabrahamy) [inline src/KafkaNetClient/Consumer.cs]: change name to "needToRefreshMetadata"

File: src/KafkaNetClient/Consumer.cs [hunk 2]
```diff
@@ -180,13 +184,14 @@ private Task ConsumeTopicPartitionAsync(string topic, int partitionId)
...
-                                        }, _disposeToken.Token);
+                                            //this is a block operation
+                                            _fetchResponseQueue.Add(message, _disposeToken.Token);
+                                        }, _disposeToken.Token).ConfigureAwait(false);
 
                                         if (_disposeToken.IsCancellationRequested) return;
                                     }
 
-                                    var nextOffset = response.Messages.Max(x => x.Meta.Offset) + 1;
+                                    var nextOffset = response.Messages[response.Messages.Count - 1].Meta.Offset + 1;
```
Comment (yoniabrahamy) [inline src/KafkaNetClient/Consumer.cs]: Why not response.Messages.Last()?

File: src/KafkaNetClient/KafkaTcpSocket.cs [hunk 3]
```diff
@@ -138,92 +142,112 @@ private Task<byte[]> EnqueueReadTask(int readSize, CancellationToken cancellatio
...
             //As long as there is one unique thread for the write operations and one unique thread for the read operations, there will be no cross-interference
             //between read and write threads and no synchronization is required.
             //https://msdn.microsoft.com/en-us/library/z2xae4f4.aspx
+
+            //Exception need to thrown immediately and not depend on the next task
+            var readTask = ProcessNetworkstreamsSendTask(netStream);
+            var sendTask = ProcessNetworkstreamTasksReadTask(netStream);
+            await Task.WhenAny(readTask, sendTask);
+            if (_disposeToken.IsCancellationRequested) return;
+            await ThrowTaskException(readTask);
```
Comment (yoniabrahamy) [inline src/KafkaNetClient/KafkaTcpSocket.cs]: Maybe change the function name to something that implies that an exception will not be necessarily thrown. Maybe ThrowTaskExceptionIfFaulted

File: src/KafkaNetClient/KafkaTcpSocket.cs [hunk 4]
```diff
@@ -389,6 +421,10 @@ private static Exception FindExceptionInLastReadOrWriteTasks(Task lastWriteTask,
                 }
 
                 await Task.Delay(TimeSpan.FromMilliseconds(reconnectionDelay), _disposeToken.Token).ConfigureAwait(false);
+                if (_maxRetry < attempts)
```
Comment (yoniabrahamy) [inline src/KafkaNetClient/KafkaTcpSocket.cs]: Move this condition inside the catch so we don't accidentally throw an exception if we succeeded.

File: src/KafkaNetClient/Producer.cs [hunk 5]
```diff
@@ -274,31 +278,57 @@ private bool IsNotDisposedOrHasMessagesToProcess()
...
-                    foreach (var task in sendTasks)
-                    {
-                        //TODO when we dont ask for an ACK, result is an empty list.  Which FirstOrDefault returns null.  Dont like this...
-                        task.MessagesSent.ForEach(async x => x.Tcs.TrySetResult(await task.Task));
-                    }
                 }
-                catch
+                catch (Exception)
+                {
+                    //TODO:Add more log
```
Comment (yoniabrahamy) [inline src/KafkaNetClient/Producer.cs]: Well then, Where is the log?

File: src/kafka-tests/Fakes/BrokerRouterProxy.cs [hunk 6]
```diff
@@ -70,7 +70,7 @@ public IBrokerRouter Create()
             });
         }
 
-        public static async Task<MetadataResponse> DefaultMetadataResponse()
+        public static async Task<MetadataResponse> CreateMetaResponseWith2Broker()
```
Comment (yoniabrahamy) [inline src/kafka-tests/Fakes/BrokerRouterProxy.cs]: The name is 2Unclear4me. Why only two brokers? what's going on?

File: src/kafka-tests/Fakes/BrokerRouterProxy.cs [hunk 7]
```diff
@@ -119,5 +119,54 @@ public IBrokerRouter Create()
                         }
                 };
         }
+
+        public static async Task<MetadataResponse> CreateMetaResponseWith1Broker()
```
Comment (yoniabrahamy) [inline src/kafka-tests/Fakes/BrokerRouterProxy.cs]: change to CreateMetadataResponseWithSingleBroker

File: src/kafka-tests/Unit/KafkaConnectionTests.cs [hunk 8]
```diff
@@ -37,7 +38,7 @@ public void Setup()
         [Test, Repeat(IntegrationConfig.NumberOfRepeat)]
         public async Task ShouldStartReadPollingOnConstruction()
         {
-            using (var socket = new KafkaTcpSocket(_log, _kafkaEndpoint))
+            using (var socket = new KafkaTcpSocket(_log, _kafkaEndpoint, 5))
```
Comment (yoniabrahamy) [inline src/kafka-tests/Unit/KafkaConnectionTests.cs]: Why not use _maxRetry?

File: src/kafka-tests/Unit/KafkaConnectionTests.cs [hunk 9]
```diff
@@ -49,7 +50,7 @@ public void Setup()
         public void ShouldReportServerUriOnConstruction()
         {
             var expectedUrl = _kafkaEndpoint;
-            using (var socket = new KafkaTcpSocket(_log, expectedUrl))
+            using (var socket = new KafkaTcpSocket(_log, expectedUrl, 5))
```
Comment (yoniabrahamy) [inline src/kafka-tests/Unit/KafkaConnectionTests.cs]: Same here..

File: src/kafka-tests/Unit/KafkaTcpSocketTests.cs [hunk 10]
```diff
@@ -36,7 +40,7 @@ public KafkaTcpSocketTests()
         [Test, Repeat(IntegrationConfig.NumberOfRepeat)]
         public void KafkaTcpSocketShouldConstruct()
         {
-            using (var test = new KafkaTcpSocket(_log, _fakeServerUrl))
+            using (var test = new KafkaTcpSocket(_log, _fakeServerUrl, 5))
```
Comment (yoniabrahamy) [inline src/kafka-tests/Unit/KafkaTcpSocketTests.cs]: Also here.

**Comment by BDeus** [comment]
hi,
i try to debug the code but i have some regression with this branch when i intentionnaly shutdown a kafka broker.
I found an infinite loop with the asynccollection and the StartReadStreamPoller.
StartReadStreamPoller add an item in the readqueue and the asynccollection remains in  DrainAndApply function (in the while dequeue loop)

**Comment by EranOfer** [comment]
hi,
Thanks alot for your help.
I will check this matter tomorrow.
If you could add a failed test it will help.

**Comment by BDeus** [comment]
i wrote a failed test and the patch of the asynccollection
do you want me to push it in your branch? or just copy and paste here?

**Comment by EranOfer** [comment]
I will glad if you do both.
Copy and paste here and create pull request on my local branch 
https://github.com/EranOfer/KafkaNetClient/tree/BugNotRecoveringAfterAnError .
I will go over the code and merge it .

**Comment by EranOfer** [comment]
and again thanks alot for your help :+1:

**Comment by randa1** [comment]
Actually, @EranOfer, I think it's better if @BDeus will create a pull request to this repository (in master branch) and not to your local one. Later you can pull the changes to your fork. It's better to handle all contributions in here than to create a multi-repository hierarchy.

**Comment by BDeus** [comment]
Finally i have create a pull request on @EranOfer branch because i had some refactoring in my branch before the patch.
I prefer than @EranOfer merge it before in is own branch.

**Pull request merged**